### PR TITLE
Buy sell

### PR DESCRIPTION
**Pull request by ordd**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

About # (link your issue here)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request ( for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 1]
```diff
@@ -27,14 +27,43 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
+                          uint256 amount,
+                          uint256 returnAmount,
+                          uint256 totalSupply,
+                          uint256 reserveBalance);
+
+    event ReserveRatioUpdated(address indexed caller,
+                              uint256 nom,
+                              uint256 denom);
+
+    event gdMintInterest(address indexed caller,
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol]: rename 
InterestMinted
UBIExpansionMinted

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 2]
```diff
@@ -54,6 +83,7 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
         onlyAvatar
     {
         reserveRatioDailyExpansion = rdiv(_nom, _denom);
+        emit ReserveRatioUpdated(msg.sender, _nom, _denom);
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol:86]: description isnt correct. you changed it so it no longer defaults to 20%
verify other methods comments are correct

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 3]
```diff
@@ -197,24 +259,25 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
         ReserveToken memory rtoken = reserveTokens[address(_token)];
         require(rtoken.gdSupply > 0, "Reserve token not initialized");
+        uint256 base = 10;
         return
             calculateSaleReturn(
                 rtoken.gdSupply,
                 rtoken.reserveSupply,
                 rtoken.reserveRatio,
-                uint256(10**gooddollar.decimals())
+                uint256(base**gooddollar.decimals())
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol]: why?

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 4]
```diff
@@ -197,24 +259,25 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
     /**
     @dev calculate how much G$ to mint based on added token supply (from interest)
     and on current reserve ratio, in order to keep G$ price the same at the bonding curve
-    @param token the reserve token
+    @param _token the reserve token
     @param _addTokenSupply amount of token added to supply
     @return how much to mint in order to keep price in bonding curve the same
      */
-    function calculateToMint(ERC20 _token, uint256 _addTokenSupply)
+    function calculateMintInterest(ERC20 _token, uint256 _addTokenSupply)
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol:308]: add formula to description

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 5]
```diff
@@ -229,12 +292,72 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
+        return toMint;
+    }
+
+    /**
+    @dev calculate how much G$ to mint based on expansion change (new reserve
+    ratio), in order to keep G$ price the same at the bonding curve
+    @param _token the reserve token
+    @return how much to mint in order to keep price in bonding curve the same
+     */
+    function calculateMintExpansion(ERC20 _token)
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol:351]: add formula to description and add inline comments during calculation

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 6]
```diff
@@ -229,12 +292,72 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
+        public
+        view
+        onlyActiveToken(_token)
+        returns (uint256)
+    {
+        ReserveToken memory reserveToken = reserveTokens[address(_token)];
+        uint32 succeedReserveRatio = calculateNewReserveRatio(_token);
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol]: rename newReserveRatio

File: stakingModel/contracts/GoodMarketMaker.sol [hunk 7]
```diff
@@ -229,12 +292,72 @@ contract GoodMarketMaker is BancorFormula, DSMath, SchemeGuard {
...
+    @param _token the reserve token
     @param _addTokenSupply amount of token added to supply
     @return how much to mint in order to keep price in bonding curve the same
      */
+    function mintInterest(ERC20 _token, uint256 _addTokenSupply)
+        public
+        onlyOwner
+        returns (uint256)
+    {
+        uint256 toMint = calculateMintInterest(_token, _addTokenSupply);
```
Comment (sirpy) [inline stakingModel/contracts/GoodMarketMaker.sol:333]: if addTokenSupply is 0 need to return 0

File: stakingModel/test/GoodMarketMaker.test.js [hunk 8]
```diff
@@ -60,16 +51,109 @@ contract(
...
+      const priceBefore = await marketMaker.currentPrice(cDAI.address);
+      await marketMaker.mintInterest(
+        cDAI.address,
+        web3.utils.numberToHex(1e8)
+      );
+      let reserveToken = await marketMaker.reserveTokens(cDAI.address);
+      let gdSupply = reserveToken.gdSupply;
+      let reserveSupply = reserveToken.reserveSupply;
+      let reserveRatio = reserveToken.reserveRatio;
+      await marketMaker.initializeToken(
```
Comment (sirpy) [inline stakingModel/test/GoodMarketMaker.test.js]: need to call mintInterest and not initialize

File: stakingModel/test/GoodMarketMaker.test.js [hunk 9]
```diff
@@ -60,16 +51,109 @@ contract(
...
+    it("should be able to calculate minted gd based on expansion of reserve ratio, the price stays the same", async () => {
+      let reserveTokenBefore = await marketMaker.reserveTokens(cDAI.address);
+      let gdSupplyBefore = reserveTokenBefore.gdSupply;
+      let reserveSupplyBefore = reserveTokenBefore.reserveSupply;
+      let reserveRatioBefore = reserveTokenBefore.reserveRatio;
+      const priceBefore = await marketMaker.currentPrice(cDAI.address);
+      const toMint = await marketMaker.calculateMintExpansion(cDAI.address);
+      const newRR = await marketMaker.calculateNewReserveRatio(cDAI.address);
+      expect(reserveRatioBefore.toString()).not.to.be.equal(newRR.toString());
+      await marketMaker.initializeToken(
```
Comment (sirpy) [inline stakingModel/test/GoodMarketMaker.test.js]: you need to call mint and not initializeToken

File: stakingModel/test/GoodMarketMaker.test.js (hunk 8 above)
Comment (sirpy) [reply stakingModel/test/GoodMarketMaker.test.js]: actually probably need to block initialiezToken for a token that is already initialized

File: stakingModel/test/GoodMarketMaker.test.js [hunk 10]
```diff
@@ -91,19 +175,223 @@ contract(
...
+      const gDReturn = await marketMaker.sellReturn(
+        cDAI.address,
+        10 //0.1 gd
+      );
+      let reserveToken = await marketMaker.reserveTokens(cDAI.address);
+      let reserveBalance = reserveToken.reserveSupply.toString();
+      let sellAmount = 10;
+      let supply = reserveToken.gdSupply.toString(); 
+      let rr = reserveToken.reserveRatio.toNumber(); 
+      const expectedReturn = reserveBalance * (1 - (1 - sellAmount / supply) ** (1000000 / rr));
```
Comment (sirpy) [inline stakingModel/test/GoodMarketMaker.test.js:181]: explain or give refernce to formula

File: stakingModel/test/GoodMarketMaker.test.js (hunk 9 above)
Comment (sirpy) [reply stakingModel/test/GoodMarketMaker.test.js]: ohh this is just a calculation... but why initialize then?

File: stakingModel/test/GoodMarketMaker.test.js [hunk 11]
```diff
@@ -91,19 +175,223 @@ contract(
...
+                      type: 'uint256',
+                      name: '_denom'
+                  }]
+              }, ['999388834642296', denom]);
+      await avatar.genericCall(marketMaker.address, encodedCall, 0);
+      let reserveRatioDailyExpansion = await marketMaker.reserveRatioDailyExpansion();
+      expect(reserveRatioDailyExpansion).not.to.be.equal(currentReserveRatioDailyExpansion);
+    });
+
+    it("should be able to buy only by the owner", async () => {
```
Comment (sirpy) [inline stakingModel/test/GoodMarketMaker.test.js]: description seems incorrect

File: stakingModel/test/GoodMarketMaker.test.js [hunk 12]
```diff
@@ -91,19 +175,223 @@ contract(
...
+          from: staker
+        }
+      );
+      const totalMinted = 1e18 / priceBefore.toNumber();
+      expect(toMint.toString()).to.be.equal((Math.floor(totalMinted * 100)).toString());
+      const priceAfter = await marketMaker.currentPrice(dai.address);
+      expect(priceBefore.toString()).to.be.equal(priceAfter.toString());
+    });
+
+    it("should not change the reserve ratio", async () => {
```
Comment (sirpy) [inline stakingModel/test/GoodMarketMaker.test.js]: description not enough, what should not change reserve ratio
same for next tests

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/test/GoodMarketMaker.test.js (hunk 9 above)
Comment (ordd) [reply stakingModel/test/GoodMarketMaker.test.js]: @sirpy 
to initialize the reserve with the new reserve ratio and the new gd supply to check the correctness of the calculation. the next test is similar, it checks the `mintExpansion` which also call to `calculateMintExpansion`.

File: stakingModel/contracts/GoodReserveCDai.sol [hunk 13]
```diff
@@ -138,9 +250,23 @@ contract GoodReserveCDai is DSMath, SchemeGuard {
...
+    function destroy(address _dest, address _newMarketMakerOwner)
+        public
+        active
+        onlyAvatar
+    {
+        uint256 cDaiBalance = cDai.balanceOf(address(this));
+        cDai.transfer(_dest, cDaiBalance);
+        require(cDai.balanceOf(address(this)) == 0, "Funds transfer has failed");
+        marketMaker.transferOwnership(_newMarketMakerOwner);
+        isActive = 0;
```
Comment (sirpy) [inline stakingModel/contracts/GoodReserveCDai.sol]: call destory?

File: stakingModel/contracts/GoodReserveCDai.sol [hunk 14]
```diff
@@ -62,40 +67,145 @@ contract GoodReserveCDai is DSMath, SchemeGuard {
...
+        active
+        onlyCDai(sellTo)
+        returns (uint256)
+    {
+        ERC20Burnable(address(gooddollar)).burnFrom(msg.sender, gdAmount);
+        uint256 deductedGdAmount = calculateSellDeduction(gdAmount);
+        uint256 tokenReturn = marketMaker.sellWithDeduction(sellTo, gdAmount, deductedGdAmount);
+        require(tokenReturn >= minReturn, "Token return must be above the minReturn");
+        require(sellTo.transfer(msg.sender, tokenReturn) == true, "Transfer failed");
+        emit TokenSold(msg.sender, address(sellTo), gdAmount, minReturn, tokenReturn);
```
Comment (sirpy) [inline stakingModel/contracts/GoodReserveCDai.sol]: maybe add the deducted amount to emitted event? call it "contribution" and not deduction
do you think minReturn should be calculated before or after deduction?

File: stakingModel/contracts/GoodReserveCDai.sol [hunk 15]
```diff
@@ -119,17 +229,19 @@ contract GoodReserveCDai is DSMath, SchemeGuard {
...
     {
         uint256 price = currentPrice(interestToken);
-        uint256 gdToMint = marketMaker.calculateToMint(interestToken, transfered);
-        uint256 precisionLoss = uint256(
-            //TODO: fix dangerous sub
-            ERC20Detailed(address(interestToken)).decimals() -
-                gooddollar.decimals()
-        );
+        uint256 gdToMint = marketMaker.mintInterest(interestToken, transfered);
+        // uint256 precisionLoss = uint256(
```
Comment (sirpy) [inline stakingModel/contracts/GoodReserveCDai.sol]: do we need these comments?

File: stakingModel/contracts/GoodReserveCDai.sol [hunk 16]
```diff
@@ -119,17 +229,19 @@ contract GoodReserveCDai is DSMath, SchemeGuard {
         uint256 interest
     )
         public
+        active
         onlyCDai(interestToken)
         onlyFundManager
         returns (uint256, uint256)
     {
         uint256 price = currentPrice(interestToken);
```
Comment (sirpy) [inline stakingModel/contracts/GoodReserveCDai.sol:240]: did you modify the currentPrice formula?

**Review by sirpy — CHANGES_REQUESTED** [review]

File: stakingModel/contracts/GoodReserveCDai.sol (hunk 14 above)
Comment (ordd) [reply stakingModel/contracts/GoodReserveCDai.sol]: I think after

File: stakingModel/contracts/GoodMarketMaker.sol (hunk 5 above)
Comment (sirpy) [reply stakingModel/contracts/GoodMarketMaker.sol:351]: waiting

File: stakingModel/contracts/GoodMarketMaker.sol (hunk 4 above)
Comment (sirpy) [reply stakingModel/contracts/GoodMarketMaker.sol:308]: waiting

**Pull request merged**